### DISCUSSION TITLE
Add value to session

### DISCUSSION THREAD

**Discussion by Psvensso** [vendurehq/vendure#858] [Q&A]
Hi, 
im working on a One-Time-Password authentication strategy and for this use-case i think it would be nice to store the OTP data on the session. Right now i can't see how i would store data on the current session object. The only way i can think of is to create my own session but that seems like a waste. If i then later add a cache strategy i would have to implement it on multiple places and... no. Cleanest would be to modify the existing session. 

Could it be an idea to have a "extra properties" object on the Session with a "SessionService > getPropterty() / setProperty()" methods? Similar idea as for Entities.

Another option i guess would be to create my own plugin and db table to store the OPT but this seems both risky and over complicated. Think it's clean that the OTP belongs to the session and feels safer when it's salted with this user's session.

Any other idea's hwo to solve the use-case?

**Comment by Psvensso** [comment]
The user has the "verified" flow with a verification token however it's not connected to the authentication method. Perhaps it's the auth method that should be verified, then if the user changes authentication method it would require a re-validation.

**Reply by michaelbromley** [reply]
As you point out, custom fields are not implemented on the Session entity. And unfortunately will not be until the next major release, as I am refraining from making breaking changes now.

However, would it not be feasible to add a custom field to the User entity? The custom field would not be cached, so to access it, you would need to perform a lookup in the DB.

> Perhaps it's the auth method that should be verified, then if the user changes authentication method it would require a re-validation.

My assumption was that in the case of external auth providers (e.g. Google, FB), we can safely assume that verification has already taken place by that provider. Does that not make sense in your case?

**Reply by Psvensso** [reply]
Hmmm... i think this can be moved to the [Make email optional on a Customer](https://github.com/vendure-ecommerce/vendure/discussions/860#discussioncomment-687297) discussion. 

However, what I mean by OTP in this context is actually when OTP is used as authentication. Not as verification. 
For example on a PhoneAuthentication flow the user would enter his/her phone and receive a OTP on each login request. 
That OTP token would be great to store in the Session since it's only valid on UnauthenticatedSessions and is not "long lived". 
But this is just one use-case and it can be solved in anyways. The expandable session object was just a "nice to have" for this usecase.

**Reply by ramkishore-yerubandi** [reply]
@Psvensso were you able to implement OTP based login for customers? Need your help on how you solved this.

**Reply by Psvensso** [reply]
@ramkishore-yerubandi no, we did not find any temporary solition or valid workaround for "OTP" style of logins. The authentication and authorization were way to close and deeply embedded. Might have change now, was some time ago but atleast as i can remember.

**Reply by ramkishore-yerubandi** [reply]
Thanks for the response @Psvensso.

**Reply by aliasifk** [reply]
@Psvensso were you able to create OTP based auth? I am doing same and stuck

**Reply by Psvensso** [reply]
As i replied to the above comment. No, username/password are/were to deeply integrated.

**Reply by michaelbromley** [reply]
As mentioned in the long thread in https://github.com/vendure-ecommerce/vendure/discussions/860, this could be accomodated as part of v2 which is currently under development. First step would be for someone to open an issue detailing the current limitations which prevent OTP flows, and suggesting changes to the APIs which would fix that.

Since discussions are not tracked like issues, I totally forgot about this whole subject. So if an issue is created I can add it to the v2.0 roadmap and them we can work towards a good solution 👍
